File: api/typescript/src/core/lib/services/port_spec.ts
import { Port } from "../../kurtosis_core_rpc_api_bindings/api_container_service_pb";

export type TransportProtocol = Port.TransportProtocol;
export namespace TransportProtocol {
    export const TCP = Port.TransportProtocol.TCP;
    export const UDP = Port.TransportProtocol.UDP;
}

// Ports are 16 bit and should be no higher than max 16-bit number
export const MAX_PORT_NUM : number = 65535;

const   allowedTransportProtocols : Set<TransportProtocol> =
    new Set<TransportProtocol>([TransportProtocol.TCP, TransportProtocol.UDP]);

export function IsValidTransportProtocol(protocol: TransportProtocol): boolean {
    return allowedTransportProtocols.has(protocol)
}

export class PortSpec {
    constructor(
        public readonly number: number,
        public readonly transportProtocol: TransportProtocol,
        public readonly maybeApplicationProtocol?: string
    ) {}
    
    // No need for getters because the fields are 'readonly'
}

File: api/golang/core/lib/services/port_spec.go
/*
 * Copyright (c) 2021 - present Kurtosis Technologies Inc.
 * All Rights Reserved.
 */

package services

import (
	"fmt"
	"github.com/kurtosis-tech/kurtosis/api/golang/core/kurtosis_core_rpc_api_bindings"
	"math"
)

// Use a type alias here to make this a bit more user-friendly
type TransportProtocol kurtosis_core_rpc_api_bindings.Port_TransportProtocol

const (
	MaxPortNum               = math.MaxUint16
	TransportProtocol_TCP    = TransportProtocol(kurtosis_core_rpc_api_bindings.Port_TCP)
	TransportProtocol_UDP    = TransportProtocol(kurtosis_core_rpc_api_bindings.Port_UDP)
	emptyApplicationProtocol = ""
)

// "Set" of allowed port protocols
var allowedTransportProtocols = map[TransportProtocol]bool{
	TransportProtocol_TCP: true,
	TransportProtocol_UDP: true,
}

func (protocol TransportProtocol) IsValid() bool {
	_, found := allowedTransportProtocols[protocol]
	return found
}

type PortSpec struct {
	number                   uint16
	transportProtocol        TransportProtocol
	maybeApplicationProtocol string
}

func NewPortSpec(number uint16, transportProtocol TransportProtocol, maybeApplicationProtocol string) *PortSpec {
	return &PortSpec{number: number, transportProtocol: transportProtocol, maybeApplicationProtocol: maybeApplicationProtocol}
}

func (spec *PortSpec) GetNumber() uint16 {
	return spec.number
}

func (spec *PortSpec) GetTransportProtocol() TransportProtocol {
	return spec.transportProtocol
}

func (spec *PortSpec) String() string {
	if spec.maybeApplicationProtocol == emptyApplicationProtocol {
		return fmt.Sprintf("%d/%v", spec.number, spec.transportProtocol)
	}
	return fmt.Sprintf("%v:%d/%v", spec.maybeApplicationProtocol, spec.number, spec.transportProtocol)
}

func (spec *PortSpec) GetMaybeApplicationProtocol() string {
	return spec.maybeApplicationProtocol
}
